plugins/pulp_rpm/plugins/distributors/yum/metadata/package.py

```diff
@@ -73,45 +69,35 @@ def add_package_group_unit_metadata(self, group_unit):
 
         if group_unit.langonly:
-            ElementTree.SubElement(group_element, 'langonly').text = group_unit.langonly
-        ElementTree.SubElement(group_element, 'name').text = group_unit.name
-        self._write_translated_fields(group_element, 'name', group_unit.translated_name)
-        ElementTree.SubElement(group_element, 'description').text = group_unit.description
-        self._write_translated_fields(group_element, 'description',
-                                      group_unit.translated_description)
-
-        package_list_element = ElementTree.SubElement(group_element, 'packagelist')
+            self.xml_generator.completeElement('langonly', {}, group_unit.langonly)
+        self.xml_generator.completeElement('name', {}, group_unit.name)
+        self._write_translated_fields('name', group_unit.translated_name)
+        self.xml_generator.completeElement('description', {}, group_unit.description)
+        self._write_translated_fields('description', group_unit.translated_description)
+
+        self.xml_generator.startElement('packagelist', {})
         if group_unit.mandatory_package_names:
             for pkg in sorted(group_unit.mandatory_package_names):
-                ElementTree.SubElement(package_list_element, 'packagereq',
-                                       {'type': 'mandatory'}).text = pkg
+                self.xml_generator.completeElement('packagereq', {'type': 'mandatory'}, pkg)
         if group_unit.default_package_names:
             for pkg in sorted(group_unit.default_package_names):
-                ElementTree.SubElement(package_list_element, 'packagereq',
-                                       {'type': 'default'}).text = pkg
+                self.xml_generator.completeElement('packagereq', {'type': 'default'}, pkg)
         if group_unit.optional_package_names:
             for pkg in sorted(group_unit.optional_package_names):
-                ElementTree.SubElement(package_list_element, 'packagereq',
-                                       {'type': 'optional'}).text = pkg
+                self.xml_generator.completeElement('packagereq', {'type': 'optional'}, pkg)
         if group_unit.conditional_package_names:
             for pkg_name, value in group_unit.conditional_package_names:
-                ElementTree.SubElement(package_list_element, 'packagereq',
-                                       {'type': 'conditional',
-                                        'requires': value}).text = pkg_name
-
-        group_element_string = ElementTree.tostring(group_element, 'utf-8')
-        _LOG.debug('Writing package_group unit metadata:\n' + group_element_string)
-        self.metadata_file_handle.write(group_element_string)
+                attrs = {'type': 'conditional', 'requires': value}
+                self.xml_generator.completeElement('packagereq', attrs, pkg_name)
+        self.xml_generator.endElement('packagelist')
+        self.xml_generator.endElement('group')
```

Comment:
I am definitely not suggesting we do this now, but offering an idea to contemplate. :) I wonder if ``startElement`` would make a nice context manager. It would call ``endElement`` at the right time based on code indentation. And the resulting code indentation would make it easier for the reader to follow where in the document they are.

```
with self.xml_generator.startElement('group', {}):
    self.xml_generator.completeElement('id', {}, group_unit.package_group_id)
    self.xml_generator.completeElement('default', {}, str(group_unit.default).lower())
    self.xml_generator.completeElement('uservisible', {}, str(group_unit.user_visible).lower())
    self.xml_generator.completeElement('display_order', {}, str(group_unit.display_order))

    if group_unit.langonly:
        self.xml_generator.completeElement('langonly', {}, group_unit.langonly)
    self.xml_generator.completeElement('name', {}, group_unit.name)
    self._write_translated_fields('name', group_unit.translated_name)
    self.xml_generator.completeElement('description', {}, group_unit.description)
    self._write_translated_fields('description', group_unit.translated_description)

    with self.xml_generator.startElement('packagelist', {}):
        if group_unit.mandatory_package_names:
            for pkg in sorted(group_unit.mandatory_package_names):
                self.xml_generator.completeElement('packagereq', {'type': 'mandatory'}, pkg)
        if group_unit.default_package_names:
            for pkg in sorted(group_unit.default_package_names):
                self.xml_generator.completeElement('packagereq', {'type': 'default'}, pkg)
        if group_unit.optional_package_names:
            for pkg in sorted(group_unit.optional_package_names):
                self.xml_generator.completeElement('packagereq', {'type': 'optional'}, pkg)
        if group_unit.conditional_package_names:
            for pkg_name, value in group_unit.conditional_package_names:
                attrs = {'type': 'conditional', 'requires': value}
                self.xml_generator.completeElement('packagereq', attrs, pkg_name)
```